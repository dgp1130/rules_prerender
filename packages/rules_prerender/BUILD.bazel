load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
load("@npm//@bazel/typescript:index.bzl", "ts_library")
load("//tools:jasmine.bzl", "jasmine_node_test")
load("//tools:publish.bzl", "publish_files")

exports_files(
    ["rollup-default.config.js"],
    visibility = ["//visibility:public"],
)

ts_library(
    name = "rules_prerender",
    srcs = ["index.ts"],
    module_name = "rules_prerender",
    visibility = ["//visibility:public"],
    deps = [
        ":scripts",
        ":styles",
        "//common/models:prerender_resource",
    ],
)

publish_files(
    name = "publish_files",
    files = glob(["**/*.bzl"]) + [
        "rollup-default.config.js",
    ],
    visibility = ["//:__pkg__"],
)

bzl_library(
    name = "entry_points",
    srcs = ["entry_points.bzl"],
)

bzl_library(
    name = "extract_single_resource",
    srcs = ["extract_single_resource.bzl"],
    visibility = ["//:__pkg__"],
)

bzl_library(
    name = "inject_resources",
    srcs = ["inject_resources.bzl"],
    visibility = ["//:__pkg__"],
)

bzl_library(
    name = "multi_inject_resources",
    srcs = ["multi_inject_resources.bzl"],
    visibility = ["//:__pkg__"],
    deps = [":web_resources"],
)

bzl_library(
    name = "prerender_component",
    srcs = ["prerender_component.bzl"],
    visibility = ["//:__pkg__"],
    deps = [
        ":web_resources",
        "//common:label",
    ],
)

bzl_library(
    name = "prerender_multi_page",
    srcs = ["prerender_multi_page.bzl"],
    visibility = ["//:__pkg__"],
    deps = [
        ":entry_points",
        ":prerender_component",
        ":web_resources",
        "//packages/renderer:build_vars",
    ],
)

bzl_library(
    name = "prerender_multi_page_bundled",
    srcs = ["prerender_multi_page_bundled.bzl"],
    visibility = ["//:__pkg__"],
    deps = [
        ":multi_inject_resources",
        ":prerender_multi_page",
        ":postcss_import_plugin_bzl",
        ":web_resources",
    ],
)

bzl_library(
    name = "prerender_page",
    srcs = ["prerender_page.bzl"],
    visibility = ["//:__pkg__"],
    deps = [
        ":entry_points",
        ":prerender_component",
        ":web_resources",
        "//packages/renderer:build_vars",
    ],
)

bzl_library(
    name = "postcss_import_plugin_bzl",
    srcs = ["postcss_import_plugin.bzl"],
)

bzl_library(
    name = "web_resources",
    srcs = ["web_resources.bzl"],
    visibility = ["//:__pkg__"],
    deps = ["//common:label"],
)

bzl_library(
    name = "web_resources_devserver",
    srcs = ["web_resources_devserver.bzl"],
    visibility = ["//:__pkg__"],
    deps = ["//common:label"],
)

ts_library(
    name = "scripts",
    srcs = ["scripts.ts"],
    deps = ["//common/models:prerender_annotation"],
)

ts_library(
    name = "scripts_test_lib",
    srcs = ["scripts_test.ts"],
    testonly = True,
    deps = [
        ":scripts",
        "@npm//@types/jasmine",
    ],
)

jasmine_node_test(
    name = "scripts_test",
    deps = [":scripts_test_lib"],
)

ts_library(
    name = "styles",
    srcs = ["styles.ts"],
    deps = ["//common/models:prerender_annotation"],
)

ts_library(
    name = "styles_test_lib",
    srcs = ["styles_test.ts"],
    testonly = True,
    deps = [
        ":styles",
        "@npm//@types/jasmine",
    ],
)

jasmine_node_test(
    name = "styles_test",
    deps = [":styles_test_lib"],
)
