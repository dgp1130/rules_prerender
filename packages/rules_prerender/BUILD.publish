exports_files(
    ["rollup-default.config.js"],
    visibility = ["//visibility:public"],
)
