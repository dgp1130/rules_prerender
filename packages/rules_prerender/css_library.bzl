load("@npm//@bazel/postcss:index.bzl", "postcss_multi_binary")
load(":postcss_import_plugin.bzl", IMPORT_PLUGIN_CONFIG = "PLUGIN_CONFIG")

_CssInfo = provider(fields = ["sources", "transitive_sources"])
CssImportMap = provider(fields = ["import_map"])

# TODO: Strict deps?
def _css_library_impl(ctx):
    return [
        DefaultInfo(files = depset(ctx.files.srcs)),
        _CssInfo(
            sources = ctx.files.srcs,
            transitive_sources = depset(ctx.files.srcs,
                transitive = [dep[_CssInfo].transitive_sources
                            for dep in ctx.attr.deps],
            ),
        ),
    ]

css_library = rule(
    implementation = _css_library_impl,
    attrs = {
        "srcs": attr.label_list(allow_files = True),
        "deps": attr.label_list(providers = [_CssInfo]),
    },
)

# It might look like this implementation should accept multiple `css_library()` targets,
# but doing so would be a bad idea. This is because the library is separated into "direct
# sources" and "transitive dependencies" which are fed into `postcss_multi_binary()`. If
# this allowed multiple libraries and same split was done, it would be "all libraries'
# direct sources" and "all libraries' transitive dependencies". This sounds reasonable,
# but would break strict deps. If one library had `@import url("rules_prerender/foo");`
# but forgot to add a dep on `//:foo` yet another library in this binary had a dep on
# `//:foo`, the first would incorrectly compile successfully. As a result, we need a
# separate `postcss_multi_binary()` for each compiled `css_library()`.
def _css_binary(name, lib, testonly = None, visibility = None):
    """TODO"""
    # Collect all transitive CSS dependencies into a single `DefaultInfo` target.
    css_transitive_deps = "%s_transitive_deps" % name
    _css_deps(
        name = css_transitive_deps,
        lib = lib,
        testonly = testonly,
    )

    # Compile the CSS. This only compiles one library, but that library may have many
    # source files, each one should be compiled into its own output, which requires
    # `postcss_multi_binary()` even though `postcss_binary()` sounds like it would be
    # sufficient here.
    # TODO: What if `lib` has two files with the same name in different directories?
    binary = "%s_postcss" % name
    postcss_multi_binary(
        name = binary,
        srcs = [lib],
        output_pattern = "{name}",
        sourcemap = True,
        plugins = {
            "//tools/internal:postcss_import_plugin": IMPORT_PLUGIN_CONFIG,
        },
        testonly = testonly,
        visibility = visibility,
        deps = [":%s" % css_transitive_deps],
    )

    # Return the binary outputs with a map of import name -> file path.
    _css_map(
        name = name,
        bin = ":%s" % binary,
        lib = lib,
    )

def css_multi_binary(name, libs, testonly = None, visibility = None):
    """TODO"""
    binaries = []
    for (index, lib) in enumerate(libs):
        binary_name = "%s_binary_%s" % (name, index)
        _css_binary(
            name = binary_name,
            lib = lib,
            testonly = testonly,
        )
        binaries.append(binary_name)
    
    css_group(
        name = name,
        deps = [":%s" % binary for binary in binaries],
        testonly = testonly,
        visibility = visibility,
    )

def _css_deps_impl(ctx):
    return DefaultInfo(
        files = ctx.attr.lib[_CssInfo].transitive_sources,
    )

_css_deps = rule(
    implementation = _css_deps_impl,
    attrs = {
        "lib": attr.label(
            mandatory = True,
            providers = [_CssInfo],
        ),
    },
)

def _css_map_impl(ctx):
    lib_files = sorted(ctx.attr.lib[_CssInfo].sources)
    bin_files = sorted([file for file in ctx.files.bin if not file.basename.endswith(".map")])

    if len(lib_files) != len(bin_files):
        fail(("Number of files from the CSS library (%s from %s) does not equal the" +
              " number of files from the CSS binary (%s from %s)") % (
                  len(lib_files),
                  ctx.attr.lib.label,
                  len(bin_files),
                  ctx.attr.bin.label,
             ))

    import_map = dict()
    lib_wksp = (ctx.attr.lib.label.workspace_name
                if ctx.attr.lib.label.workspace_name
                else ctx.workspace_name)
    for index in range(len(lib_files)):
        lib_file = lib_files[index]
        bin_file = bin_files[index]

        # The library and binary file should have the same name, but might be in different
        # packages.
        if lib_file.basename != bin_file.basename:
            fail(("CSS library files did not match up with CSS binary files.\n" +
                  "Lib files:\n%s\n\nBin files:\n%s") % (
                      "\n".join([file.path for file in lib_files]),
                      "\n".join([file.path for file in bin_files]),
                  ))

        key = "%s/%s" % (lib_wksp, lib_file.short_path)
        if key in import_map:
            fail("CSS library file (%s) mapped twice, once to %s and a second time to %s." % (
                key,
                import_map[key].path,
                bin_file.path,
            ))

        import_map[key] = bin_file

    return [
        ctx.attr.bin[DefaultInfo],
        CssImportMap(import_map = import_map)
    ]

_css_map = rule(
    implementation = _css_map_impl,
    attrs = {
        "bin": attr.label(mandatory = True),
        "lib": attr.label(
            mandatory = True,
            providers = [_CssInfo],
        ),
    },
)

def _css_group_impl(ctx):
    return [
        DefaultInfo(
            files = depset([], transitive = [dep[DefaultInfo].files
                                             for dep in ctx.attr.deps]),
        ),
        CssImportMap(
            import_map = _merge_import_maps([dep[CssImportMap]
                                             for dep in ctx.attr.deps
                                             if CssImportMap in dep]),
        ),
    ]

css_group = rule(
    implementation = _css_group_impl,
    attrs = {
        "deps": attr.label_list(mandatory = True),
    },
)

def _merge_import_maps(css_import_maps):
    """Merges a list of `CssImportMap` into a single `CssImportMap`.
    
    Fails the build if the same import path appears as a key in two maps.
    """
    import_map = dict()
    for css_import_map in css_import_maps:
        for (key, value) in css_import_map.import_map.items():
            if key in import_map:
                fail(("Found duplicate CSS import path in `_css_group()`. %s maps to" +
                      " both %s and %s." % (key, import_map[key], value)))
            import_map[key] = value

    return import_map
