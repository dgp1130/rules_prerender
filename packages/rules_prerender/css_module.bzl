load("@bazel_skylib//rules:build_test.bzl", "build_test")
load("@npm//@bazel/postcss:index.bzl", "postcss_binary")
load("@npm//@bazel/typescript:index.bzl", "ts_library")
load("//common:label.bzl", "absolute")
load(":postcss_import_plugin.bzl", IMPORT_PLUGIN_CONFIG = "PLUGIN_CONFIG")
load(":postcss_modules_plugin.bzl", MODULES_PLUGIN_CONFIG = "PLUGIN_CONFIG")

# TODO: bzl_library()
def css_module(name, srcs, testonly = None, visibility = None, deps = []):
    """Compiles the given CSS sources as CSS modules.

    Definition of CSS modules: https://github.com/css-modules/css-modules.

    Rerpresents a "library" of CSS files which can depend on other CSS files or
    be included in a `prerender_*()` target. Note that this rule does not
    actually have any outputs which are considered public API. The only
    supported way of consuming this rule is to depend on it with a
    `prerender_*()` target.
    
    Params:
        name: The name of this target.
        srcs: CSS source files to compile as CSS modules.
        testonly: https://docs.bazel.build/versions/master/be/common-definitions.html#common.testonly:~:text=testonly,-Boolean%3B
        visibility: https://docs.bazel.build/versions/master/be/common-definitions.html#common.visibility:~:text=use.-,visibility
        deps: Other `css_module()` targets depended upon by this target via an
            `@import` in one of the source files.
    
    Outputs:
        `%{name}`: INTERNAL - A `ts_library()` which provides a `default export`
            object that maps each source class name to its locally scoped name.
        `%{name}_modules`: INTERNAL - A `filegroup()` of all the compiled CSS
            module files.
        `%{name}_build_test`: INTERNAL - A test to confirm that all the relevant
            targets build successfully.
    """
    sources = "%s_sources" % name
    module_deps = ["%s_modules" % absolute(dep) for dep in deps]

    # Convert each CSS file into a TypeScript wrapper that exports its classes.
    # TODO: Use `postcss_mutli_binary()` when it supports `additional_outputs`.
    # https://github.com/bazelbuild/rules_postcss/issues/64
    [postcss_binary(
        name = "%s_%s_css_module" % (name, ".".join(src.split(".")[:-1])),
        src = src,
        output_name = _module_name(src),
        additional_outputs = ["%s.ts" % src],
        sourcemap = True,
        plugins = {
            "//tools/internal:postcss_modules_plugin": MODULES_PLUGIN_CONFIG,
        },
        testonly = testonly,
        deps = [
            ":%s" % sources,
            "@npm//css-modules-loader-core",
        ],
    ) for src in srcs]

    # Compile the CSS module TypeScript wrappers.
    ts_library(
        name = name,
        srcs = ["%s.ts" % src for src in srcs],
        testonly = testonly,
        visibility = visibility,
    )

    # Export transitive sources.
    native.filegroup(
        name = sources,
        srcs = srcs + ["%s_sources" % dep for dep in deps],
        testonly = testonly,
        visibility = visibility,
    )

    # Export transitive source files.
    native.filegroup(
        name = "%s_modules" % name,
        srcs = [_module_name(src) for src in srcs] + module_deps,
        testonly = testonly,
        visibility = visibility,
    )

    # TODO: Strict deps check, but don't use `@import` plugin by default, or
    # else styles will be duplicated for each dependency.
    build_test(
        name = "%s_build_test" % name,
        targets = [
            ":%s" % name,
            ":%s_modules" % name,
        ],
    )

def _module_name(src):
    return "%s.module.css" % ".".join(src.split(".")[:-1])
