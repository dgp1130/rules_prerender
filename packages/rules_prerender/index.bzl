"""Re-exports public Starlark symbols."""

load(":inject_resources.bzl", _inject_resources = "inject_resources")
load(":prerender_component.bzl", _prerender_component = "prerender_component")
load(":prerender_page.bzl", _prerender_page = "prerender_page")

inject_resources = _inject_resources
prerender_component = _prerender_component
prerender_page = _prerender_page
