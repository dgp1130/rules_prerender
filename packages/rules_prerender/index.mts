/** @fileoverview Re-exports public symbols. */

export {
    includeScript,
    includeScriptAnnotation as internalIncludeScriptAnnotation,
} from './scripts.mjs';
export {
    inlineStyle,
    inlineStyleAnnotation as internalInlineStyleAnnotation,
    InlineStyleNotFoundError as InternalInlineStyleNotFoundError,
} from './styles.mjs';

export {
    setMap as internalSetInlineStyleMap,
    resetMapForTesting as internalResetInlineStyleMapForTesting,
} from './inline_style_map.mjs';
export {
    execrootRelative as internalExecrootRelative,
    parseExecrootRelativePath as internalSplitExecrootRelativePath,
} from './paths.mjs';

export {
    FileSystem as InternalFileSystem,
    diskFs as internalDiskFs,
} from '../../common/file_system.mjs';
export { PrerenderResource } from '../../common/models/prerender_resource.mjs';
export { SafeHtml, isSafeHtml } from '../../common/safe_html/safe_html.mjs';
export { unsafeTreatStringAsSafeHtml } from '../../common/safe_html/unsafe_html.mjs';
