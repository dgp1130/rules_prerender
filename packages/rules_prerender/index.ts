/** @fileoverview Re-exports public symbols. */

export { PrerenderResource } from '../../common/models/prerender_resource';
export { includeScript } from './scripts';
export { inlineStyle, InlineStyleNotFoundError as InternalInlineStyleNotFoundError } from './styles';

export { setMap as internalSetInlineStyleMap, resetMapForTesting as internalResetInlineStyleMapForTesting } from './inline_style_map';
