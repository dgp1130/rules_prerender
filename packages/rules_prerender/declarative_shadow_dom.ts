import { includeScript } from 'rules_prerender';

/** TODO: Document. */
export function polyfillDeclarativeShadowDom(): string {
    return includeScript('rules_prerender/packages/rules_prerender/declarative_shadow_dom_polyfill');
}
