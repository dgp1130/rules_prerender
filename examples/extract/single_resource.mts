import { PrerenderResource } from 'rules_prerender';

/** Generates a single resource. */
export default function*(): Generator<PrerenderResource, void, void> {
    yield PrerenderResource.fromText('/foo.txt', 'I\'m a single resource!');
}
