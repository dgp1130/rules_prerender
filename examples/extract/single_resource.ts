import { PrerenderResource } from 'rules_prerender';

/** Generates a single resource. */
export default function*(): Iterable<PrerenderResource> {
    yield PrerenderResource.of(`/foo.txt`, `I'm a single resource!`);
}
