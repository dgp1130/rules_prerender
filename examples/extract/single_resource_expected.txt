I'm a single resource!